lib/Adapter/Tolerant/TolerantIndexBuilder.php

```diff
@@ -0,0 +1,155 @@
+<?php
+
+namespace Phpactor\Indexer\Adapter\Tolerant;
+
+use Microsoft\PhpParser\Node;
+use Microsoft\PhpParser\Node\QualifiedName;
+use Microsoft\PhpParser\Node\Statement\ClassDeclaration;
+use Microsoft\PhpParser\Node\Statement\FunctionDeclaration;
+use Microsoft\PhpParser\Parser;
+use Phpactor\Indexer\Model\Index;
+use Phpactor\Indexer\Model\IndexBuilder;
+use Phpactor\Indexer\Model\Record\ClassRecord;
+use Phpactor\Indexer\Model\Record\FunctionRecord;
+use Phpactor\Name\FullyQualifiedName;
+use Phpactor\TextDocument\ByteOffset;
+use SplFileInfo;
+
+class TolerantIndexBuilder implements IndexBuilder
+{
+    /**
+     * @var Index
+     */
+    private $index;
+
+    /**
+     * @var Parser
+     */
+    private $parser;
+
+    public function __construct(
+        Index $index,
+        ?Parser $parser = null
+    ) {
+        $this->index = $index;
+        $this->parser = $parser ?: new Parser();
+    }
+
+    public function index(SplFileInfo $info): void
+    {
+        $contents = @file_get_contents($info->getPathname());
+
+        if (false === $contents) {
+            return;
+        }
+
+        $node = $this->parser->parseSourceFile($contents, $info->getPathname());
+        $this->indexNode($info, $node);
+    }
+
+    public function done(): void
+    {
+        $this->index->updateTimestamp();
+    }
+
+    private function indexNode(SplFileInfo $info, Node $node): void
+    {
+        if ($node instanceof ClassDeclaration) {
+            $this->indexClassDeclaration($info, $node);
+            return;
+        }
+
+        if ($node instanceof FunctionDeclaration) {
+            $this->indexFunction($info, $node);
+            return;
+        }
+
+        foreach ($node->getChildNodes() as $childNode) {
+            $this->indexNode($info, $childNode);
+        }
+    }
+
+    private function indexClassDeclaration(SplFileInfo $info, ClassDeclaration $node): void
+    {
+        $record = $this->index->get(ClassRecord::fromName($node->getNamespacedName()->getFullyQualifiedNameText()));
+        assert($record instanceof ClassRecord);
```

Comment:
Is this for real necessity or to infer the type ?
In case of the later I would suggest `/** @var ClassRecord $record */` instead to remove one assertion per node, well class node but still :)
I'm willing to say that if the indexer does not return the proper record we have bigger issues so it might be tolerable to let the script failed by itself if it does not have the proper record.
Which will be unit tested at least anyway.

Reply:
`assert` cost nothing in "prod env" if properly setup. :}

Reply:
It's just to help with completion -- I really want to add generics support soon :D

Reply:
> assert cost nothing in "prod env" if properly setup. :}

Didn't knew that!
Feel kind of strange, some reading to do :)